Rewrite contribution tools documentation (#2203)
carbon-language/carbon-lang: docs/project/contribution_tools.md

## Table of contents

-   [Setup commands](#setup-commands)
    -   [Debian or Ubuntu](#debian-or-ubuntu)
    -   [macOS](#macos)
-   [Tools](#tools)
    -   [Main tools](#main-tools)
        -   [Running pre-commit](#running-pre-commit)
    -   [Optional tools](#optional-tools)
    -   [Manually building Clang and LLVM (not recommended)](#manually-building-clang-and-llvm-not-recommended)
-   [Troubleshooting build issues](#troubleshooting-build-issues)
    -   [How LLVM is found](#how-llvm-is-found)
    -   [Old LLVM versions](#old-llvm-versions)
    -   [Asking for help](#asking-for-help)
-   [Troubleshooting debug issues](#troubleshooting-debug-issues)

<!-- tocstop -->

## Setup commands

These commands should help set up a development environment on your machine.

### Debian or Ubuntu

```
# Update apt.
sudo apt-get update

# Install apt tools.
sudo apt-get install \
  bazel \
  clang \
  gh \
  libc++-dev \
  lld \
  python3 \

>   zlib1g-dev

Reviewer: ```suggestion
sudo apt update

# Install tools.
sudo apt install \
  bazel \
  clang \
  gh \
  libc++-dev \
  lld \
  python3 \
  zlib1g-dev
```

# Install pre-commit.
pip3 install pre-commit

# Set up git.

> gh repository fork --clone carbon-language/carbon-lang

Reviewer: I think this is not the right command if you already have a fork of the carbon project (also doesn't seem like per-machine setup, but I think it still makes sense to list here). Maybe:

```suggestion
# If you don't already have a fork:
gh repo fork --clone carbon-language/carbon-lang
# Otherwise:
gh repo clone carbon-lang
```

(At least on my machine, `gh repository` doesn't work and you need to use `gh repo`.)

Author: `repository` was an incorrect autofix, fixing that.

`gh repo fork --clone carbon-language/carbon-lang` should always work (it detects already-created forks). I consider giving one command versus a choice of commands a matter of simplicity.

Reviewer: Makes sense. I didn't see that mentioned in `gh help repo fork` but I've tested it and it does seem to work as you describe.

cd carbon-lang
pre-commit install

# Run tests.
bazel test //...:all
```

> NOTE: Most LLVM 14+ installs should build Carbon. If you're having issues, see
> [troubleshooting build issues](#troubleshooting-build-issues).

### macOS

```
# Install Hombrew.
/bin/bash -c "$(curl -fsSL \
  https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

# IMPORTANT: Make sure `brew` is added to the PATH!

# Install Homebrew tools.
brew install \
  gh \
  llvm \
  python@3.10

# IMPORTANT: Make sure `llvm` is added to the PATH! It's separate from `brew`.

# Install pre-commit.
pip3 install pre-commit

# Set up git.

> gh repository fork --clone carbon-language/carbon-lang

Reviewer: ```suggestion
# If you don't already have a fork:
gh repo fork --clone carbon-language/carbon-lang
# Otherwise:
gh repo clone carbon-lang
```

Author: Same as the other response

cd carbon-lang
pre-commit install

# Run tests.
bazel test //...:all

> ```

Reviewer: I somewhat wonder if we should separate the "install pre-requisites" step which varies between systems from the "install pre-commit, set up a checkout, run a build" step which is the same across systems.

Author: My sense is the amount of duplication is pretty low, and handy to just have one spot to look at. Also, we have 2 systems of a popular 3; Windows would probably be different.

# Run tests.

> bazel test //...:all

Reviewer: ```suggestion
bazelisk test //...:all
```
You mention this below, should we be recommending it here?

Author: `bazel` should work with the brew install (it symlinks bazel to bazelisk), and has the nice aspect of being a consistent command. Where do I mention running `bazelisk` directly?

> ```
> # For `brew`, `gh`, and other tools:
> export PATH="${HOME}/.brew/bin:${PATH}"
> # For `llvm`:
> export PATH="$(brew --prefix llvm)/bin:${PATH}"
> ```

## Tools

### Main tools

These tools are essential for work on Carbon.

-   Package managers
    -   `apt` (for Debian or Ubuntu)
        -   To upgrade versions of `apt` packages, it will be necessary to

>             periodically run `sudo apt-get update && sudo apt-get upgrade`.

Reviewer: ```suggestion
            periodically run `sudo apt update && sudo apt upgrade`.
```

    -   [Homebrew](https://brew.sh/) (for macOS)
        -   To upgrade versions of `brew` packages, it will be necessary to
            periodically run `brew upgrade`.
    -   [`python3` and `pip3`](https://python.org)
        -   Carbon requires Python 3.9 or newer.
        -   To upgrade versions of `pip3` packages, it will be necessary to
            periodically run `pip3 list --outdated`, then
            `pip3 install -U <package>` to upgrade desired packages.
        -   When upgrading, version dependencies may mean packages _should_ be
            outdated, and not be upgraded.

> -   Main tools

Reviewer: Prettier is not listed here; should it be?

Author: I was avoiding listing things run by `pre-commit`.

    -   [Bazel](https://www.bazel.build/)
    -   [Bazelisk](https://docs.bazel.build/versions/master/install-bazelisk.html)

>         (for macOS): Handles Bazel versions.

Reviewer: The instructions above don't say to install or use `bazelisk`.

Author: Fixed

Reviewer: The instructions above still don't say to use `bazelisk`; I've added a suggestion where I would have expected to see this used.

Incidentally, why is this only listed as "(for macOS)"?

Author: > The instructions above still don't say to use `bazelisk`; I've added a suggestion where I would have expected to see this used.

See above response -- brew symlinks bazel to bazelisk

> Incidentally, why is this only listed as "(for macOS)"?

Because apt seemed easier for installing bazel on Linux, even though it's not as consistent. Alternatives are installing bazelisk via brew or `go get`, and both felt like poor choices (`brew` being a lot of work for one command, `go get` requiring more work and requiring non-standard manual updates)

Reviewer: OK. It looks strange to say to install `bazelisk` and to mention it here, but never say to run it anywhere, but I guess that's just how this works with `brew`. Fair enough.

    -   [Clang](https://clang.llvm.org/) and [LLVM](https://llvm.org/)
        -   NOTE: Most LLVM 14+ installs should build Carbon. If you're having
            issues, see
            [troubleshooting build issues](#troubleshooting-build-issues).
    -   [gh CLI](https://github.com/cli/cli): Helps with GitHub.
    -   [pre-commit](https://pre-commit.com): Validates and cleans up git
        commits.
-   Libraries
    -   zlib1g-dev: Used as a library, but not installed on all Linux systems.

Our recommended way of installing is:

```bash
brew install bazelisk

> ```

Reviewer: This doesn't really seem to fit here. Should this be moved elsewhere or deleted?

Author: Fixed

#### Running pre-commit

[pre-commit](https://pre-commit.com) is typically set up using
`pre-commit install`. When set up in this mode, it will check for issues when

> `git commit` is run. A typical commit workflow looks like:

Reviewer: I don't know that we should be recommending this. Prettier will trash .md files that have certain Markdown errors in them (eg, unterminated \`\`\` blocks), meaning that `git commit -a` can do damage that's hard to undo. At least when working on .md files, I would strongly recommend installing as a pre-push hook instead of a pre-commit hook for that reason. But I think using a pre-push hook is better advice in general, given that it frees up `git commit` for local incremental development milestones with reformatting only forced when publishing changes.

Author: I've added a mention that some people prefer to use it pre-push. Does that work for you?

While I understand your perspective, I've tried pre-push and don't like it. I'd rather find out I have bad content earlier. It's also consistent with https://pre-commit.com/#usage, and I think it's helpful to avoid diverging from intended ways to run tools unless necessary.

Reviewer: Yeah, I can live with this :)

2. `git add .` to add the automatically modifications done by `pre-commit`.
3. `git commit` again.

You can also use `pre-commit run` to check pending changes without `git commit`,
or `pre-commit run -a` to run on all files in the repository.

> NOTE: Some developers prefer to run `pre-commit` on `git push` instead of
> `git commit` because they want to commit files as originally authored instead
> of with pre-commit modifications. To switch, run
> `pre-commit uninstall && pre-commit install -t pre-push`.

### Optional tools

These tools aren't necessary to contribute to Carbon, but can be worth
considering if they fit your workflow.

-   [GitHub Desktop](https://desktop.github.com/): A UI for managing GitHub
    repositories.
-   `rs-git-fsmonitor` and Watchman: Helps make `git` run faster on large
    repositories.
    -   **WARNING**: Bugs in `rs-git-fsmonitor` and/or Watchman can result in
        `pre-commit` deleting files. If you see files being deleted, disable
        `rs-git-fsmonitor` with `git config --unset core.fsmonitor`.
-   [vim-prettier](https://github.com/prettier/vim-prettier): A vim integration

>     for Prettier, which we use for formatting.

Reviewer: If we're not going to list Prettier itself above, should we do so here? I assume the `vim` plugin depends on it.

Author: I don't think Prettier needs to be installed separately, see: https://github.com/prettier/vim-prettier#prettier-executable-resolution

            you should be prompted to use Carbon's
            [devcontainer](/.devcontainer/Dockerfile) with "Reopen in
            container".

### Manually building Clang and LLVM (not recommended)

We primarily test against <apt.llvm.org> and Homebrew installations. However,
you can build and install LLVM yourself if you feel more comfortable with it.
The essential CMake options to pass in order for this to work reliably include:

```
-DLLVM_ENABLE_PROJECTS=clang;clang-tools-extra;lld
-DLLVM_ENABLE_RUNTIMES=compiler-rt;libcxx;libcxxabi;libunwind
-DRUNTIMES_CMAKE_ARGS=-DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF;-DCMAKE_POSITION_INDEPENDENT_CODE=ON;-DLIBCXX_ENABLE_STATIC_ABI_LIBRARY=ON;-DLIBCXX_STATICALLY_LINK_ABI_IN_SHARED_LIBRARY=OFF;-DLIBCXX_STATICALLY_LINK_ABI_IN_STATIC_LIBRARY=ON;-DLIBCXX_USE_COMPILER_RT=ON;-DLIBCXXABI_USE_COMPILER_RT=ON;-DLIBCXXABI_USE_LLVM_UNWINDER=ON
```

## Troubleshooting build issues

### How LLVM is found

The [clang detection](/bazel/cc_toolchains/clang_configuration.bzl) decides
where to use LLVM from by first checking `$CC` then `which clang`. It will use

> other LLVM tools in the same directory of the `clang` that it finds.

Reviewer: Is this for the `clang` we use as a compiler, or for the LLVM library that we compile and link against? (Do we require them to be the same? Do we pick up an `llvm-config` from adjacent to the `clang` binary to find the LLVM libraries?)

Author: Note, I'm removing this section. Your questions suggest to me that it may create more questions than it addresses, and we didn't have it before.

> Is this for the clang we use as a compiler, or for the LLVM library that we compile and link against?

The compiler. The LLVM library comes from the WORKSPACE.

> Do we require them to be the same?

They literally cannot be, as we cannot compile the code from source without a compiler.

> Do we pick up an llvm-config from adjacent to the clang binary to find the LLVM libraries?)

No, we don't use llvm-config (kind of looks like it's more for if you were trying to build with gcc?)

Reviewer: > > Do we require them to be the same?
> 
> They literally cannot be, as we cannot compile the code from source without a compiler.

They can be the same llvm/clang revision, and with bootstrapping they can be built from literally the same source code.

> > Do we pick up an llvm-config from adjacent to the clang binary to find the LLVM libraries?)
> 
> No, we don't use llvm-config (kind of looks like it's more for if you were trying to build with gcc?)

`llvm-config` is the way you're supposed to find which flags you pass to your compiler when you want to use the LLVM libraries, including `-I`, `-L`, `-D`, etc. It's not specific to using gcc or clang or icc or whatever.

Given that we're using LLVM from the WORKSPACE, though, we know exactly where it is (and we're building it ourselves from source anyway) so this doesn't matter.
